# Дашборд отображающий динамику и структуру трудоустройства

Для благотворительного фонда Рауль необходимо подготовить дашборд отображающий динамику и структуру трудоустройства кандидатов.

Требования к дашборду:

- Параметры динамики:
    - помесячно,
    - поквартально,
    - ежегодно.
    
- Параметры структуры:
    - принадлежность благополучателя к одному из сегментов,
    - ответственная НКО,
    - регион,
    - пол благополучателя,
    - возраст,
    - тип трудоустройства (временное, постоянное с сопровождением, постоянное без сопровождения),
    - тип работодателя (внутренняя защищённая площадка, внешняя защищённая площадка, открытый рынок).
    
- Параметры трудоустройства: от 1 дня, от 1 месяца

Импортирую необходимые библиотеки:

In [ ]:
import pandas as pd
from io import BytesIO
import requests
import datetime

Прочитаю данные из файла на гугл диске:

In [ ]:
spreadsheet_id = ''
file_name = 'https://docs.google.com/spreadsheets/d/{}/export?format=csv'.format(spreadsheet_id)
r = requests.get(file_name)
data = pd.read_csv(BytesIO(r.content))

Посмотрю общую информацию о массиве:

In [ ]:
data.head(10)

Unnamed: 0  id_кандидата      пол  возраст     дата_добавления  \
0           0             1  Женский       25  06.12.2018 0:00:00   
1           1             2  Мужской       28  22.09.2021 0:00:00   
2           2             3  Женский       31  08.07.2016 0:00:00   
3           3             4  Мужской       34  07.10.2022 0:00:00   
4           4             4  Мужской       34  07.10.2022 0:00:00   
5           5             4  Мужской       34  07.10.2022 0:00:00   
6           6             5  Мужской       16  01.03.2023 0:00:00   
7           7             6  Женский       28  01.08.2016 0:00:00   
8           8             7  Женский       33  21.12.2020 0:00:00   
9           9             8  Мужской       21  17.06.2022 0:00:00   

  инвалидность сиротский_опыт адаптированная_образовательная_программа  \
0       ИСТИНА           ЛОЖЬ                                     ЛОЖЬ   
1       ИСТИНА           ЛОЖЬ                                     ЛОЖЬ   
2       ИСТИНА           ЛОЖЬ                                     ЛОЖЬ   
3       ИСТИНА           ЛОЖЬ                                     ЛОЖЬ   
4       ИСТИНА           ЛОЖЬ                                     ЛОЖЬ   
5       ИСТИНА           ЛОЖЬ                                     ЛОЖЬ   
6         ЛОЖЬ         ИСТИНА                                     ЛОЖЬ   
7       ИСТИНА           ЛОЖЬ                                     ЛОЖЬ   
8       ИСТИНА           ЛОЖЬ                                   ИСТИНА   
9       ИСТИНА         ИСТИНА                                   ИСТИНА   

  начало_трудоустройства завершение_трудоустройства 1+_месяц_работы  \
0     16.01.2023 0:00:00                        NaN          ИСТИНА   
1     09.01.2023 0:00:00                        NaN          ИСТИНА   
2     18.04.2023 0:00:00         30.06.2023 0:00:00          ИСТИНА   
3     21.11.2022 0:00:00         30.12.2022 0:00:00          ИСТИНА   
4     22.03.2023 0:00:00         30.03.2023 0:00:00            ЛОЖЬ   
5     31.03.2023 0:00:00                        NaN          ИСТИНА   
6     28.04.2023 0:00:00         15.06.2023 0:00:00          ИСТИНА   
7     28.11.2022 0:00:00                        NaN          ИСТИНА   
8     01.03.2023 0:00:00         31.05.2023 0:00:00          ИСТИНА   
9     09.03.2023 0:00:00                        NaN          ИСТИНА   

  тип_площадки               тип_трудоустройства тип_опыта_работы регион_код  \
0      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
1      Внешняя   Открытый рынок с сопровождением        Повторный     reg200   
2      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
3      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
4      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
5      Внешняя  Открытый рынок без сопровождения        Повторный     reg200   
6      Внешняя   Открытый рынок с сопровождением           Первый     reg200   
7      Внешняя         Временное трудоустройство        Повторный     reg200   
8      Внешняя   Открытый рынок с сопровождением        Повторный     reg200   
9      Внешняя   Открытый рынок с сопровождением        Повторный     reg200   

  провайдер_код  
0      prov3500  
1      prov3500  
2      prov3500  
3      prov3500  
4      prov3500  
5      prov3500  
6      prov3500  
7      prov3500  
8      prov3500  
9      prov3500

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 259 entries, 0 to 258
Data columns (total 16 columns):
 #   Column                                    Non-Null Count  Dtype 
---  ------                                    --------------  ----- 
 0   Unnamed: 0                                259 non-null    int64 
 1   id_кандидата                              259 non-null    int64 
 2   пол                                       259 non-null    object
 3   возраст                                   259 non-null    int64 
 4   дата_добавления                           259 non-null    object
 5   инвалидность                              259 non-null    object
 6   сиротский_опыт                            259 non-null    object
 7   адаптированная_образовательная_программа  259 non-null    object
 8   начало_трудоустройства                    259 non-null    object
 9   завершение_трудоустройства                85 non-null     object
 10  1+_месяц_работы                           259 non-

**Типы данных** в некоторых столбцах не соответствуют содержимому, преобразую их:

In [ ]:
columns = ['инвалидность','сиротский_опыт','адаптированная_образовательная_программа','1+_месяц_работы']

for column in columns:
    data[column] = data[column].replace(['ИСТИНА','ЛОЖЬ'],[True,False])


In [ ]:
data[['дата_добавления', 'начало_трудоустройства','завершение_трудоустройства']] = \
data[['дата_добавления', 'начало_трудоустройства','завершение_трудоустройства']].apply(pd.to_datetime,
                                                                                       format='%d.%m.%Y %H:%M:%S')

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 259 entries, 0 to 258
Data columns (total 16 columns):
 #   Column                                    Non-Null Count  Dtype         
---  ------                                    --------------  -----         
 0   Unnamed: 0                                259 non-null    int64         
 1   id_кандидата                              259 non-null    int64         
 2   пол                                       259 non-null    object        
 3   возраст                                   259 non-null    int64         
 4   дата_добавления                           259 non-null    datetime64[ns]
 5   инвалидность                              259 non-null    bool          
 6   сиротский_опыт                            259 non-null    bool          
 7   адаптированная_образовательная_программа  259 non-null    bool          
 8   начало_трудоустройства                    259 non-null    datetime64[ns]
 9   завершение_трудоустройства      

Проверю данные на **дубликаты**:

In [ ]:
print('Дубликатов в таблице data:',data.duplicated().sum())

Дубликатов в таблице data: 0


Проверю массив на **пропуски**:

In [ ]:
data.isna().sum()

Unnamed: 0                                    0
id_кандидата                                  0
пол                                           0
возраст                                       0
дата_добавления                               0
инвалидность                                  0
сиротский_опыт                                0
адаптированная_образовательная_программа      0
начало_трудоустройства                        0
завершение_трудоустройства                  174
1+_месяц_работы                               0
тип_площадки                                  1
тип_трудоустройства                           1
тип_опыта_работы                              1
регион_код                                    0
провайдер_код                                 0
dtype: int64

Очевидно что трудоустройство может быть не завершено - эти пропуски заполнять нет смысла, остальные пропуски заполнить не представляется возможным, однако необходимо известить заказчика о них. Посмотрю детальнее на строки с пропусками:

In [ ]:
display(data.loc[data['дата_добавления'].isna() == True])
display(data.loc[data['тип_площадки'].isna() == True])
display(data.loc[data['тип_трудоустройства'].isna() == True])
display(data.loc[data['тип_опыта_работы'].isna() == True])

Empty DataFrame
Columns: [Unnamed: 0, id_кандидата, пол, возраст, дата_добавления, инвалидность, сиротский_опыт, адаптированная_образовательная_программа, начало_трудоустройства, завершение_трудоустройства, 1+_месяц_работы, тип_площадки, тип_трудоустройства, тип_опыта_работы, регион_код, провайдер_код]
Index: []

Unnamed: 0  id_кандидата      пол  возраст дата_добавления  инвалидность  \
209         209           197  Женский       19      2021-05-01         False   

     сиротский_опыт  адаптированная_образовательная_программа  \
209            True                                     False   

    начало_трудоустройства завершение_трудоустройства  1+_месяц_работы  \
209             2023-08-01                        NaT             True   

    тип_площадки тип_трудоустройства тип_опыта_работы регион_код провайдер_код  
209          NaN                 NaN              NaN     reg204      prov3504

Unnamed: 0  id_кандидата      пол  возраст дата_добавления  инвалидность  \
209         209           197  Женский       19      2021-05-01         False   

     сиротский_опыт  адаптированная_образовательная_программа  \
209            True                                     False   

    начало_трудоустройства завершение_трудоустройства  1+_месяц_работы  \
209             2023-08-01                        NaT             True   

    тип_площадки тип_трудоустройства тип_опыта_работы регион_код провайдер_код  
209          NaN                 NaN              NaN     reg204      prov3504

Unnamed: 0  id_кандидата      пол  возраст дата_добавления  инвалидность  \
209         209           197  Женский       19      2021-05-01         False   

     сиротский_опыт  адаптированная_образовательная_программа  \
209            True                                     False   

    начало_трудоустройства завершение_трудоустройства  1+_месяц_работы  \
209             2023-08-01                        NaT             True   

    тип_площадки тип_трудоустройства тип_опыта_работы регион_код провайдер_код  
209          NaN                 NaN              NaN     reg204      prov3504

В итоге имеем неполные данные для кандидата с id = 213 - отсутствует дата добавления и для кандидата с id = 197 отстутствуют данные о типе площадки, типе трудоустройства и типе опыта работы.

Оценю общую адекватность данных, посмотрю максимальные и минимальные значения, характер распределения данных:

In [ ]:
data.describe()

Unnamed: 0  id_кандидата     возраст
count  259.000000    259.000000  259.000000
mean   129.000000    122.293436   22.359073
std     74.911058     70.736428    5.574585
min      0.000000      1.000000   14.000000
25%     64.500000     61.500000   18.000000
50%    129.000000    124.000000   22.000000
75%    193.500000    182.500000   25.000000
max    258.000000    242.000000   39.000000

Аномальных значений не обнаружено, для построения имеем данные по трудоустройсту с ноября 2022 по август 2023 года,возраст кандидатов от 14 до 39 лет.  

Количество уникальных кандидатов:

In [ ]:
len(data.id_кандидата.unique())

242

**Вывод:**

Для анализа предоставлен массив данных в  259 строк с информацией о трудоустройстве 242 уникальных кандидатов.'
   
Дубликатов не обнаружено.
   
Проверка на пропуски показала, что дата завершения трудоустройства отстуствует в 174 строках, что допустимо, поскольку кандидаты могут не завершить трудоустройство.

Помимо этого имеем неполные данные для кандидата с id = 213 - отсутствует дата добавления и для кандидата с id = 197 отстутствуют данные о типе площадки, типе трудоустройства и типе опыта работы - эти данные исказят результаты дашборда, поскольку данные кандидаты будут исключены при фильтрации по соответствующим признакам.

Аномальных значений не обнаружено, для построения дашборда имеем данные по трудоустройсту с ноября 2022 по август 2023 года,возраст кандидатов от 14 до 39 лет.  

Поскольку заказчик четко определил задачу для дашборда как ответ на вопрос - сколько уникальных кандидатов было трудоустроено в зависимостки различных параметров, - дополнительных исследований данных не требуется.  

Данные подготовлены для дальнейшего построения дашборда.

Ссылка на подготовленный дашборд: https://datalens.yandex/p7gcus2p83uwe